### ISSUE TITLE
Suport for Somfy Opening Detector io

### ISSUE THREAD

**Issue by AdyRock** [totalimmersion/com.somfy.tahoma#17]
I have only just got Homey so I hope this is the correct place to put this. Could you add support for the Opening Detector io?
I have one on my patio door and I want to open the curtains when the door opens and close them after a short delay after the door is closed. Tahoma can do it apart from the delay.

Thank you for the consideration,
Adrian

**Comment by totalimmersion** [comment]
Hi Adrian,

I think it would be possible to add support for the opening detector, however I don't own one, therefore you need to help me with some information about this device. I've created a document describing the steps you need to take in order to get the information needed.

Best,

Lennart

Ps. Since 99% of all the Homey owners is dutch, I'm guessing you're dutch as well. Therefore if you like we can continue this converstion in dutch.

[Instructions for the TaHoma devices.pdf](https://github.com/totalimmersion/com.somfy.tahoma/files/2089039/Instructions.for.the.TaHoma.devices.pdf)

**Comment by AdyRock** [comment]
Hi Lennart,

Thanks for looking into this. The requested information:
```javascript
{
  "creationTime": 1522277410000,
  "lastUpdateTime": 1522277410000,
  "label": "PatioDoor",
  "shortcut": false,
  "controllableName": "io:SomfyContactIOSystemSensor",
  "definition": {
    "commands": [

    ],
    "states": [
      {
        "values": [
          "closed",
          "open"
        ],
        "type": "DiscreteState",
        "qualifiedName": "core:ContactState"
      },
      {
        "type": "ContinuousState",
        "qualifiedName": "core:RSSILevelState"
      },
      {
        "values": [
          "dead",
          "lowBattery",
          "maintenanceRequired",
          "noDefect"
        ],
        "type": "DiscreteState",
        "qualifiedName": "core:SensorDefectState"
      },
      {
        "values": [
          "available",
          "unavailable"
        ],
        "type": "DiscreteState",
        "qualifiedName": "core:StatusState"
      }
    ],
    "dataProperties": [

    ],
    "widgetName": "ContactSensor",
    "uiClass": "ContactSensor",
    "qualifiedName": "io:SomfyContactIOSystemSensor",
    "type": "SENSOR"
  },
  "states": [
    {
      "name": "core:ContactState",
      "type": 3,
      "value": "closed"
    },
    {
      "name": "core:StatusState",
      "type": 3,
      "value": "available"
    },
    {
      "name": "core:RSSILevelState",
      "type": 2,
      "value": 100.0
    }
  ],
  "attributes": [
    {
      "name": "core:PowerSourceType",
      "type": 3,
      "value": "battery"
    }
  ],
  "available": true,
  "enabled": true,
  "widget": "ContactSensor",
  "type": 2,
  "uiClass": "ContactSensor"
},
```

If you need any further information I am very happy to help.

Ps. Sorry, I only speak English.

Thanks,
Adrian.

**Comment by totalimmersion** [comment]
Hi Adrian,

Thank you for your response. No worries, English is just fine. I think I'm able to support your device. However, I've got 2 more questions. First of all, I have no idea what the product looks like and I would like to add an image of the product to the Homey app store. Could you therefore point me to actual product on the website of Somfy?
Finally, I need to figure out how the status of the sensor is being updated in Tahoma. Could you tell me if there is a separate screen in Tahoma (for example if you click on the icon of the sensor) where you can view the status. If so, I would like to get to know the url and response that is being loaded if you click on that icon. Would that be possible?

Best,

Lennart

**Comment by AdyRock** [comment]
Hi Lennart,

The link to the product is: https://shop.somfy.co.uk/opening-detector-io

**Comment by totalimmersion** [comment]
Hi Adrian,

Thank you, that was exactly what I was looking for. I think this is all I need. I'll let you when it's finished or if I have any further questions.

Best,

Lennart

**Comment by paulvanderheijden90** [comment]
Hoi Lennart,
Meer support van IO devices zou voor mij de doorslag geven om een homey te kopen.
Ik heb mijn huis volledig voorzien van Somfy rolluiken. Somfy scenario knoppen Somfy io deur contacten zo als hier word beschreven.
Ik zou dus bijv savonds bij zons ondergang de rolluiken dicht sturen ( dit kan nu al met Somfy ) maar de rolluiken 10% open laten als ergens een raam open staat. Mocht de io sensor van Somfy worden toegevoegd weet ik zeker dat Homey dat kan. Heb je een idee of hier meer support over komt

**Comment by totalimmersion** [comment]
Hoi Paul,

Ja, het ziet er goed uit met het toevoegen van de openingsmelder. Dit is bijna klaar.

Groet,

Lennart

**Comment by totalimmersion** [comment]
Hi Adrian,

I've managed to add the opening detector to the app. I only need to add some icons and images, but technically it should work. Since I don't own an opening detector, I'm not able to test it. I don't know how tech savvy you are, but would you be able to test it? You can install it locally by downloading the source code from github: https://github.com/totalimmersion/com.somfy.tahoma/archive/develop.zip and follow the steps described here: https://developer.athom.com/docs/apps/tutorial-Getting%20Started.html

If this sounds too difficult, no worries, I will then finish it and release it as a beta version in the app store, however it would be best to make sure everything works as expected before releasing it.

Best,

Lennart

**Comment by paulvanderheijden90** [comment]
Zou geweldig zijn! Wat ga je nog meer toevoegen

Verstuurd vanaf mijn iPhone

**Comment by totalimmersion** [comment]
Wat ik nog meer ga toevoegen kan ik nog niet zeggen. Dat hangt helemaal af van Homey community. Ik ben volledig afhankelijk van anderen voor het toevoegen van devices die ik zelf niet bezit. Anderen zullen een request moeten indienen en de nodige informatie moeten geven aan de hand waarvan ik dan nieuwe apparaten kan toevoegen.

**Comment by paulvanderheijden90** [comment]
Interessant 

Als de io opener goed werkt ga ik de Homey aanschaffen. Daarna wil ik je graag helpen



Verstuurd vanaf mijn iPhone

**Comment by AdyRock** [comment]
Hi Lennart,

I have run the debug version and it allows me to add the sensor but there are some issues.

1. Can add the status information on the mobile card to show if the door is open or closed.

2. The flow triggers are not working correctly. The Alarm cards never trigger and all the state cards fire once every minute regardless of the door state.

3. Can you add a tag to "The state of the opening device has changed" card to show the new state. Then it can be added to logic in the ...and... column to select the then or else.

I have attached the debug output for you. The "Original" file is using the code as you left it. The "Extra" file contains some extra information form two log commands I have added. (I have altered the numbers that look to be sensitive data)

[Tahoma Log Extra.txt](https://github.com/totalimmersion/com.somfy.tahoma/files/2145554/Tahoma.Log.Extra.txt)
[Tahoma Log Original.txt](https://github.com/totalimmersion/com.somfy.tahoma/files/2145555/Tahoma.Log.Original.txt)

Let me know if you want me to try anything else.

Regards,
Adrian.

**Comment by totalimmersion** [comment]
Thanks a lot Adrian, this really helps. I'll have a look at this.

Best,

Lennart

**Comment by AdyRock** [comment]
Hi Lennart,

I couldn't resist investigating further. It all seems to be working now.
There were a couple of things: = instead of == in driver.js on one of the conditions, alarm_contact is predefined as a Boolean so change comparisons to xxx == true or false.
Thanks for your help and hope this is ok.
[com.somfy.tahoma-develop_ady.zip](https://github.com/totalimmersion/com.somfy.tahoma/files/2146365/com.somfy.tahoma-develop_ady.zip)

The attached zip contains the files I have changed

**Comment by paulvanderheijden90** [comment]
Nice to follow the hard work very interested 
Thank you i hope it Will work very soon
Verstuurd vanaf mijn iPhone

**Comment by totalimmersion** [comment]
Hi Adrian,

I've taken a look at your changes and have implemented them, so that alarm_contact is now a boolean. I've refactored the code a bit and also changed the token in such a way that it is a boolean as well instead of a string. I hope this all works (I still don't have any means to test it). Could you therefore once more run it on your system and check if everything works as expected?

Best,

Lennart

**Comment by AdyRock** [comment]
Hi Lennart,

The new version is looking good. The only thing I am not sure about is if you need the "The state of the opening detector is open / closed" in either the When or And cards. Those cards don't appear to work in the When and do the same as "The contact alarm is on" or off via the invert box in the And. So I think removing them would make it less confusing without any loss in function?

The only other request is could the poll time be a setting somewhere. I have currently reduced it to 5 seconds so I don't have to wait so long for the state to be registered. Hopefully Somfy will add some sort of push notification when they release the API. The Tahoma box responds instantly to the change in state so it presumably gets a push notification from the sensor. So for now I use Tahoma to open the curtains when the door opens and Homey to close them after a delay when the door closes.

I have attached an icon.svg file if you want to use it.

[icon.zip](https://github.com/totalimmersion/com.somfy.tahoma/files/2177682/icon.zip)

Thank you for adding this in, you have done a great job.

**Comment by totalimmersion** [comment]
Hi Adrian,

I see what you mean. I think the "The contact alarm is on" cards are automatically added by the alarm_contact capability (because I didn't add them). Therefore the other ones are not neccessary. I've removed the 2 cards for both the 'When' and 'And' section. For now I've changed the interval to 5 seconds, in the feature I will try to make this a setting the user can adjust. I've also added the icon. Thank you for supplying it. Could you once more confirm everything still works as expected? If it does, I'll finalize this version and submit it to Athom.

Best,

Lennart

**Comment by AdyRock** [comment]
Hi Lennart,
It is all working very well now.
Thank you.

**Comment by paulvanderheijden90** [comment]
Nice work Guys ! Nice to follow the hard work. So is iT now availibile  for Everyone ? 

 

Verstuurd vanaf mijn iPhone

**Comment by totalimmersion** [comment]
Hi Paul,

I've just submitted the app to Athom, so I expect the update is publicly available within a couple of days.

Best,

Lennart

**Comment by totalimmersion** [comment]
Hi Adrian,

Thank you for your help. I've also managed to make the sync interval a preference which can be changed by the user as you suggested. This will be available in the update as well.

Best,

Lennart

**Comment by totalimmersion** [comment]
Hey guys,

I just wanted to let you know that version 1.4.0 of the app has been released in the app store. Your Homey device should be automatically updated to this version. I'm not getting any error reports, so all looks good. I'm closing this issue now. If you have any new issues or new devices that you would like to have added, please create a new issue and I will do my best to handle it as quickly as possible. Adrian, thank you for your input and support, that was really helpful.

Best,

Lennart

**Comment by paulvanderheijden90** [comment]
Hallo Lennart,

Graag zou ik de homey aanschaffen. "Helaas/Gelukkig” heb ik alles van somfy. Nu zie ik dat jij bezig ben met het ontwikkelen van de app. 

Ik heb jou en Adrian gevolgd tijdens het ontwikkelen van de IO Deur melder. ( Geweldig !)

Ik heb op dit moment:
- Rolluiken Somfy IO   	( Heb je al ) 
- Deur Contacten IO 	( Heb je al ) 
- Rookmelders IO 		
- IO Sunis Wirefree II io, zonsensor  ( Heb je al ) 
- IO senario player. 
- Hitachi Airco op IO 

Hoe kijk jij er tegen aan om de andere componenten ook te ontwikkelen ? 


Met vriendelijke groet,


Paul van der Heijden

**Comment by AdyRock** [comment]
Hi Lennart,

 

Just to let you know I have been using your app on Homey V2 and it is working great. The only observations are:

1.       It can’t be installed afresh because the "compatibility": "1.x >=1.5.0", would need to be "compatibility": ">=1.5.0". As it happens I had it installed before updating Homey so it continued to work for me.

2.       The svg icon I supplied looks like a pin prick so I have added the viewport information to fix that and the new file is attached. 

 

I have installed the app with the above amendments via the CLI and the IO opening detector and the Sunis sun sensor (io) are working perfectly for me but obviously I can’t test the other devices as I don’t have them.

 

Best regards,

Adrian
